beacon-chain/sync/backfill/backfill.md

```diff
@@ -0,0 +1,853 @@
+
+#### Non-Retryable Errors
+- Chain broken (parent_root mismatch after multiple retries)
+- Configuration errors
+- Context cancellation
+
+## State Machine
+
+### Batch States
+
+```
+batchNil            → Zero value, not yet initialized
+batchInit           → Initialized, ready to be scheduled
+batchSequenced      → Scheduled to workers
+batchErrRetryable   → Failed with retryable error
+batchBlobSync       → Downloading blobs
+batchImportable     → Ready to import to DB
+batchImportComplete → Successfully imported
+batchEndSequence    → Signals end of backfill
+```
+
+### State Transitions
+
+```
+┌──────────────┐
+│  batchNil    │
+└──────┬───────┘
+       │ Initialize bounds
+       ↓
+┌──────────────┐
+│  batchInit   │◄──────────────────┐
+└──────┬───────┘                   │
+       │ sequence()                │ Retry after delay
+       ↓                           │
+┌──────────────┐                   │
+│batchSequenced│                   │
+└──────┬───────┘                   │
+       │ Worker downloads blocks   │
+       ↓                           │
+  ┌─────────────────────┐          │
+  │ Verification        │          │
+  └────┬────────────────┘          │
+       │                           │
+       ├─Success (no blobs)────────┼──────┐
+       │                           │      │
+       ├─Success (blobs needed)─┐  │      │
+       │                        ↓  │      │
+       │                ┌──────────────┐  │
+       │                │ batchBlobSync│  │
+       │                └──────┬───────┘  │
+       │                       │          │
+       │                Download blobs    │
+       │                       │          │
+       │                       ↓          │
+       │                   Success?       │
+       │                    /    \        │
+       │                  Yes    No       │
+       │                   │      │       │
+       ├───────────────────┘      │       │
+       │                          │       │
+       │◄─────────────────────────┘       │
+       │ (batchErrRetryable)              │
+       │                                  │
+       ↓                                  │
+┌──────────────┐                          │
+│batchImportable│◄────────────────────────┘
+└──────┬───────┘
+       │ Import to DB
+       ↓
+┌──────────────┐
+│batchImport   │
+│  Complete    │
+└──────────────┘
+```
+
+### End Condition
+
+When `batcher.before(upTo)` is called with `upTo <= min`:
+```go
+return batch{begin: upTo, end: upTo, state: batchEndSequence}
+```
+
+When all batches are `batchEndSequence`:
+- Worker pool's `complete()` returns `errEndSequence`
+- Service marks completion and exits
+
+## Database Integration
+
+### Schema
+
+**Backfill Status**:
+- **Bucket**: `blocksBucket`
+- **Key**: `backfillStatusKey`
+- **Value**: Marshaled `BackfillStatus` protobuf
+
+**Blocks**:
+- Saved via `SaveROBlocks(ctx, blocks, cache=false)`
+- No caching during backfill (blocks are historical)
```

Comment:
In case you want to say more about this - we cache recent blocks in the db layer for faster lookups eg replays. Since backfill blocks are not useful on those fast access paths, we added this param to the save method specifically for backfill, to avoid replacing recent/useful blocks in the cache with old/less-useful backfill blocks.